### ISSUE TITLE
Improve loading time

### ISSUE THREAD

**Issue by ThomasLeigh** [SYtor/Kanji-Dojo#195]
Forcing the app close didn't help.

**Comment by ThomasLeigh** [comment]
It shows the main screen after few minutes. But then I forced closing the app and it acts the same.

**Comment by ThomasLeigh** [comment]
It also repeats when I tap on the main screen icon.

**Comment by syt0r** [comment]
Was it an infinite loading or just a really long loading? Because it seems like there's the same issue with infinite loading as before: #135 and #119. I've managed to reproduce it by reinstalling and relaunching the app many times on an old version of android. Unfortunately the error message doesn't make any sense so I can't be certain that my fix will help but I'm going to include and test it in the next release

![Screenshot 2024-09-19 at 13 28 52 (2)](https://github.com/user-attachments/assets/2a163009-c614-4f82-8abe-5f5abb16b19f)

**Comment by neithernut** [comment]
I can's speak for @ThomasLeigh, but I experience long loading time (often more than 10s, but under a minute) on a Fairphone 3+ running /e/OS 2.3 (Android 13).

I planned to look at the code in the future, but my suspicion is that collecting info for the 'Daily Streak' section takes most of the time.

**Comment by syt0r** [comment]
Can anyone share their backup file? Maybe you have tons more of reviews compared to me. On my device loading is barely visible

**Comment by ThomasLeigh** [comment]
> Can anyone share their backup file?

Yes. I also have the impression that this has something to do with Daily Streak calculation... because even if the app is running, sometimes when I tap on the first bottom tab, it is loading long again.

It's not endless loading, it's tangibly long loading.

I will try to attach the backup...

**Comment by ThomasLeigh** [comment]
Here You are my current backup file:

[kanji-dojo-backup-2024-9-19-22-8.zip](https://github.com/user-attachments/files/17065987/kanji-dojo-backup-2024-9-19-22-8.zip)

**Comment by ThomasLeigh** [comment]
@syt0r With ver. 2.1.3 the long loading issue persists. It occurs not only while launching the app, but also often while evoking Letters or Home screen (e.g., after a finished practice), which is a tangible burden.

**Comment by ThomasLeigh** [comment]
@syt0r Any chance for eliminating this long loading issue?

**Comment by o-zig** [comment]
I've had this issue since version 2.1.2 and I played around in the app if I could somehow solve it myself. In learning/reviewing Kanji I use JLPT, WaniKani and the grades, though I am keeping the different systems in separate decks.

I made a backup before reviewing my last deck today and tested the following:
 - As usual, all separate decks I do the review: ~10 seconds processing time after the review
 - Combining all of my decks into one massive deck, then reviewing: ~2 seconds processing time after the review.

Some guesses as to the cause would be:
 - Duplicates of the same Kanji in different decks
 - Many and or large decks

Excuse my chaos, but you can find my backup attached:
[kanji-dojo-backup-2024-12-22-22-43.zip](https://github.com/user-attachments/files/18223343/kanji-dojo-backup-2024-12-22-22-43.zip)

**Comment by syt0r** [comment]
As I remember from the last time I investigated this there's already a caching mechanism there but performance may have degraded because cards are now sorted by their expected review date. I'll take a look if the logic there can by multithreaded if it's not yet like this and if there's not too much to do it might be included in the next version

**Comment by syt0r** [comment]
From new investigation seems like 90~% of loading time is spent on reading data for each card. Multithreading makes it worse as is because accessing db is the bottleneck so it needs some sort of caching to to avoid having drastic UX changes. It takes around 2.3 seconds on my device using debug build to load data with 107k review history from one of the backups above and only 0.341s on my PC, I wonder how you guys get 10s+

**Comment by ThomasLeigh** [comment]
@syt0r  No idea, I can screen record this if You want... Just let me know.

**Comment by ThomasLeigh** [comment]
@syt0r

I've observed that the issue occurs when the app has been thrown away from the recent apps (which usually happens automatically after a few hours of not using it - in my case during nights). This is how it looks the next morning:

https://github.com/user-attachments/assets/f4bf46dc-0269-40d5-a4a9-e2e70b4f937d

While after an hour (or so) of not-using it:

https://github.com/user-attachments/assets/d772cec3-f91e-432f-8885-a95178264f89

Interestingly, if I leave the app with Kanji search result screen and just hide the app (instead of going back to its home screen and Android desktop, which seems to close the app), the app remains opened even during the night.

One could change Android preferences regarding optimizing the battery and Kanji Dojo, this would be considered a workaround for some users, but other users may not prefer to use their battery / RAM for such purposes, they would prefer Kanji Dojo be independent of that and manage to fast load regardless.
